src/psyclone/psyad/transformations/preprocess.py

```diff
@@ -75,3 +81,87 @@ def preprocess_trans(kernel_psyir):
         elif oper.operator == BinaryOperation.Operator.MATMUL:
             # Apply MATMUL transformation
             matmul_trans.apply(oper)
+
+    # Deal with any associativity issues here as AssignmentTrans
+    # is not able to.
+    for assignment in kernel_psyir.walk(Assignment):
+        associativity(assignment, active_variable_names)
+
+
+def associativity(assignment, active_variable_names):
+    '''Repeatedly look for the patterns x * (a +- b) or (a +- b) */ x on
+    the rhs of this assignment where x is an inactive expression and a
+    and b are active expressions, replacing these patterns with x*a +-
+    x*b and a*/x +- b*/x respectively.
+
+    :param assignment: the Assignment Node that we are looking at.
+    :type assignment: :py:class:`psyclone.psyir.nodes.Assignment`
+    :param active_variable_names: list of active variable names.
+    :type active_variable_names: list of str
+
+    '''
+    if node_is_active_names(assignment.rhs, active_variable_names):
```

Comment:
As discussed on Teams, please could you investigate whether expanding `core/symbolic_maths.py` to support `expand` results in any simplification?

Reply:
Thanks for the pointer and the discussion on how to implement such a function. This works well, apart from failing with array notation. I've therefore added a new expand function based on sympy but kept the old one until support for array notation is added. I've also added an issue for the problem with array notation, see #1655.